### ISSUE TITLE
Unable to use keras.application within TF Estimator

### ISSUE THREAD

**Issue by sgasse** [tensorflow/tensorflow#25670]
I am having problems using keras.applications within an estimator. After much trying out myself, I came up with a minimal version to indicate the problem. I presume that it is caused by keras only having one session at a time while estimators set up several sessions for training and evaluation.

This is my [original post from stackoverflow](https://stackoverflow.com/questions/54589386/tensorflow-integrate-keras-model-in-estimator-model-fn)

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, see below
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.12.0
- Python version: 3.6.7

**Describe the current behavior**
The estimator apparently does not learn the data, which can be observed by a constant loss (in evaluation) and a stagnant loss during training.

**Describe the expected behavior**
The estimator should learn the MNIST dataset at least as good from the model features as from the raw data.

**Code to reproduce the issue**
```
import tensorflow as tf
import numpy as np

from keras.datasets import mnist


# switch to example 1/2/3
EXAMPLE_CASE = 1

# flag for initial weights loading of keras model
_W_INIT = True


def dense_net(features, labels, mode, params):
    # --- code to load a keras application ---

    # commenting in this line leads to a bump in the loss everytime the
    # evaluation is run, this indicating that keras does not handle well the
    # two sessions of the estimator API
    # tf.keras.backend.set_learning_phase(mode == tf.estimator.ModeKeys.TRAIN)
    global _W_INIT

    model = tf.keras.applications.MobileNet(
        input_tensor=features,
        input_shape=(128, 128, 3),
        include_top=False,
        pooling='avg',
        weights='imagenet' if _W_INIT else None)

    # only initialize weights once
    if _W_INIT:
        _W_INIT = False

    # switch cases
    if EXAMPLE_CASE == 1:
        # model.output is the same as model.layers[-1].output
        img = model.layers[-1].output
    elif EXAMPLE_CASE == 2:
        img = model(features)
    elif EXAMPLE_CASE == 3:
        # do not use keras features
        img = tf.keras.layers.Flatten()(features)
    else:
        raise NotImplementedError

    # --- regular code from here on ---
    for units in params['dense_layers']:
        img = tf.keras.layers.Dense(units=units, activation='relu')(img)

    logits = tf.keras.layers.Dense(units=10,
                                   activation='relu')(img)

    # compute predictions
    probs = tf.nn.softmax(logits)
    predicted_classes = tf.argmax(probs, 1)

    # compute loss
    loss = tf.losses.sparse_softmax_cross_entropy(labels, logits)

    acc = tf.metrics.accuracy(labels, predicted_classes)
    metrics = {'accuracy': acc}
    tf.summary.scalar('accuarcy', acc[1])

    if mode == tf.estimator.ModeKeys.EVAL:
        return tf.estimator.EstimatorSpec(
            mode, loss=loss, eval_metric_ops=metrics)

    # create training operation
    assert mode == tf.estimator.ModeKeys.TRAIN

    optimizer = tf.train.AdagradOptimizer(learning_rate=0.01)
    train_op = optimizer.minimize(loss, global_step=tf.train.get_global_step())
    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)


def prepare_dataset(in_tuple, n):
    feats = in_tuple[0][:n, :, :]
    labels = in_tuple[1][:n]
    feats = feats.astype(np.float32)
    feats /= 255
    labels = labels.astype(np.int32)
    return (feats, labels)


def _parse_func(features, labels):
    feats = tf.expand_dims(features, -1)
    feats = tf.image.grayscale_to_rgb(feats)
    feats = tf.image.resize_images(feats, (128, 128))
    return (feats, labels)


def load_mnist(n_train=10000, n_test=3000):
    train, test = mnist.load_data()
    train = prepare_dataset(train, n_train)
    test = prepare_dataset(test, n_test)
    return train, test


def train_input_fn(imgs, labels, batch_size):
    dataset = tf.data.Dataset.from_tensor_slices((imgs, labels))
    dataset = dataset.map(_parse_func)
    dataset = dataset.shuffle(500)
    dataset = dataset.repeat().batch(batch_size)
    return dataset


def eval_input_fn(imgs, labels, batch_size):
    dataset = tf.data.Dataset.from_tensor_slices((imgs, labels))
    dataset = dataset.map(_parse_func)
    dataset = dataset.batch(batch_size)
    return dataset


def main(m_dir=None):
    # fetch data
    (x_train, y_train), (x_test, y_test) = load_mnist()

    train_spec = tf.estimator.TrainSpec(
        input_fn=lambda: train_input_fn(
            x_train, y_train, 30),
        max_steps=150)

    eval_spec = tf.estimator.EvalSpec(
        input_fn=lambda: eval_input_fn(
            x_test, y_test, 30),
        steps=100,
        start_delay_secs=0,
        throttle_secs=0)

    run_cfg = tf.estimator.RunConfig(
        model_dir=m_dir,
        tf_random_seed=2,
        save_summary_steps=2,
        save_checkpoints_steps=10,
        keep_checkpoint_max=1)

    # build network
    classifier = tf.estimator.Estimator(
        model_fn=dense_net,
        params={
            'dense_layers': [256]},
        config=run_cfg)

    # fit the model
    tf.estimator.train_and_evaluate(
        classifier,
        train_spec,
        eval_spec)


if __name__ == "__main__":
    main()
```

**Other info / logs**

**Comment by jvishnuvardhan** [comment]
This is not Build/Installation or Bug/Performance issue. Please post this kind of support questions at [Stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow). There is a big community to support and learn from your questions. GitHub is mainly for addressing bugs in installation and performance. I have seen your clear problem description on Stackoverflow and I think community will respond to your question soon. Thanks!

**Comment by dksb** [comment]
Are you satisfied with the resolution of your issue?
[Yes](https://goo.gl/forms/Oe0tEvODFRoI2gJF3)
[No](https://goo.gl/forms/fUjzOfrtkFbrOT8d2)

**Comment by cgarciae** [comment]
I think this should be a bug, how can it be that you cannot train an estimator using a Keras application? This is a super common use-case. We've had to resort to saving the Keras application as a saved model and loading it as a predictor in the input_fn, to deploy we've had to merge the two saved models together in a separate script.

All this doesn't make sense. If TFHub enabled you to have variable image input side all would be fine, but it doesn't, that leaves keras as the only source for easy-to-use pretrained models, but if you can fine tune them using the estimator then the high-level API of Tensorflow is broken.

**Comment by nomis-g** [comment]
I agree with @cgarciae since it is said everywhere that Keras and Tensorflow work hand in hand - though when using the estimator function `train_and_evaluate`, it apparently breaks.

It would be good to have an official answer from somebody who fully understands the inner workings. E.g. I can only assume that a problem arises from Keras only having one session while `train_and_evaluate` (as far as I know) creates a training session in the beginning and a new evaluation session on each evaluation.

So in my opinion, this should be considered a bug. Fine-tuning the layers of a pretrained model is a very common use-case. And if not a functionality bug, then it should at least be considered a documentation issue: The documentation should state clearly that you (probably) cannot use pretrained keras models in the estimator.

@jvishnuvardhan can you at least confirm/explain how Keras and Tensorflow Estimators handle the sessions if used together?

**Comment by wirtsi** [comment]
I would like to add that this seems to be not only wonky for keras.application models but also for regular keras models. I have a functional keras model that trains wonderfully through model.fit but when using estimators, it never converges.

What struck me weird is that also the metrics for the estimator seem totally wrong. I am attaching metrics to the keras model plus some on the EvalSpecs of the estimator

So when using tensorflow.metrics on the estimator, I get

```
eg.
def eval_metrics(features, labels, predictions):
    return {
        'streaming_true_positives': metrics.true_positives(labels=labels, predictions=predictions['final']),
        'streaming_true_negatives': metrics.true_negatives(labels=labels, predictions=predictions['final']),
        'streaming_false_positives': metrics.false_positives(labels=labels, predictions=predictions['final']),
        'streaming_false_negatives': metrics.false_negatives(labels=labels, predictions=predictions['final'])  
    }

streaming_false_negatives = 0.0
streaming_false_positives = 304213.0
streaming_true_negatives = 0.0
streaming_true_positives = 3883.0
```

as result for the eval set, but my batch wise keras metrics give me
eg.
```
def true_positives(y_true, y_pred):
  true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
  return true_positives

false_negatives = 1.4632323
false_positives = 0.0756128
true_negatives = 126.31117
true_positives = 0.14997923
```
-> These are mean values per batch

So the estimator completely fails to work with the data somehow.

**Comment by jvishnuvardhan** [comment]
@wirtsi Could you provide any code to reproduce the bug? Thanks!

**Comment by wirtsi** [comment]
Hah, I got it ... that took me really long to figure out because at first my sample case was working. But I was able to reproduce this! The estimator metrics completely freak out if there is a Keras Embedding in the Graph, it appears as if the model never learns.

My test case is here: https://colab.research.google.com/drive/194Puigi-LdzxZup6LNREk47l9uP0_Dx9

So when training the model with keras model.fit() it works as expected

> Epoch 1/100
>  - 4s - loss: 0.6918 - acc: 0.5367 - true_positives: 146.4400 - true_negatives: 121.8800 - false_positives: 128.0600 - false_negatives: 103.5600 - recall: 0.5859 - precision: 0.5361 - val_loss: 0.6925 - val_acc: 0.5250 - val_true_positives: 130.9000 - val_true_negatives: 131.6000 - val_false_positives: 116.1000 - val_false_negatives: 121.4000 - val_recall: 0.5190 - val_precision: 0.5300
> ...
> Epoch 100/100
>  - 3s - loss: 0.0040 - acc: 1.0000 - true_positives: 249.9600 - true_negatives: 249.9800 - false_positives: 0.0000e+00 - false_negatives: 0.0000e+00 - recall: 1.0000 - precision: 1.0000 - **val_loss: 0.1620 - val_acc: 0.9492 - val_true_positives: 237.6000 - val_true_negatives: 237.0000 - val_false_positives: 11.5000 - val_false_negatives: 13.9000 - val_recall: 0.9447 - val_precision: 0.9539**

So it learned the case alright. Now when resetting the workspace and running the estimator training instead of model.fit, I get

> INFO:tensorflow:Saving dict for global step 162: binary_accuracy = 0.7645, false_negatives = 55.25, false_negatives_streaming = 0.0, false_positives = 62.5, false_positives_streaming = 992.0, global_step = 162, loss = 0.52764004, precision = 0.7601666, precision_streaming = 0.504, recall = 0.78157234, recall_streaming = 1.0, true_negatives = 185.5, true_negatives_streaming = 0.0, true_positives = 196.75, true_positives_streaming = 1008.0
> ...
> (({'binary_accuracy': 0.9345, 'false_negatives': 21.25, **'false_negatives_streaming': 0.0**, 'false_positives': 11.5, '**false_positives_streaming': 992.0,** 'loss': 0.17386678, 'precision': 0.9526271, **'precision_streaming': 0.504**, 'recall': 0.915865, **'recall_streaming': 1.0,** 'true_negatives': 236.5, **'true_negatives_streaming': 0.0**, 'true_positives': 230.75, **'true_positives_streaming': 1008.0**, 'global_step': 2000}, [])

So in fact the model seems to train alright, but the eval_metrics (from the estimator) are messed up. According to the *_streaming metrics, all the model ever does is predict true for everything, even though that doesn't seem to be the case.

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by nomis-g** [comment]
I cannot speak for @wirtsi but the fact that you cannot seamlessly use `keras.applications` inside a `tf.estimator` remains a problem

**Comment by wirtsi** [comment]
I have been digging around this for a while, it seems to be a general problem with using Keras models in Tensorflow. I have documented my findings here: 

https://stackoverflow.com/questions/54910215/tensorflow-estimator-fails-to-converge-on-model-converted-from-keras-when-using

In my case that means: Keras is totally happy to use a single neuron output layer for a binary classification task. If you use this model (which works in Keras!) and give it to an Estimator, it will not converge but only predict trues. You _must_ use two output neurons for this to work.

I can only guess that this might have to do with different ways of backpropagation but at that depth level, I am lost ....

So without wanting to hijack this issue, I would assume that this might be a similiar problem.

**Comment by Jasonnor** [comment]
I see guide of tf2.0 [here](https://www.tensorflow.org/alpha/guide/migration_guide#using_a_custom_model_fn) about using custom model_fn with keras model. 
Don't know if it can help you, but there are some code snippets that lack detailed explanations.

**Comment by Jasonnor** [comment]
Update: Seems like even using code same as document above, keras application still can not use within estimator. Testing environment: tf 1.12 and 1.13, 2.0 not tested.

**Comment by sdll** [comment]
Any updates on this?

**Comment by cmcneil** [comment]
As far as I can tell, this issue persists in TF2 while using a custom training loop. If I define a MobilenetV2 Keras model myself (by copying the author's code), then I can train it just fine, but if I use tf.keras.applications.MobileNetV2(weights=None), it simply ignores the gradient tape, and I can't update the weights of the network at all.

**Comment by cgarciae** [comment]
The problem in TF 1.x was related to `tf.keras` having a global session different to the one used by the estimator. Given that sessions are no longer a thing in 2.0 I am surprised its not working @cmcneil, you should report this as a separate bug.

**Comment by frozenzo** [comment]
Any update on this? It seems that the problem still reserves now in TF2.3. @cgarciae @sgasse  @cmcneil

**Comment by golden0080** [comment]
Any update on this issue? I'm using TF1 and seeing training issues whenever enabled Keras in estimator.

**Comment by rmothukuru** [comment]
@sgasse,
Can you please confirm if this issue is still relevant in **`Tensorflow 2.x Version`**, as we predominantly use [Keras Layers](https://www.tensorflow.org/api_docs/python/tf/keras/layers) instead of [Estimators](https://www.tensorflow.org/guide/estimator) and that [tf.keras.applications](https://www.tensorflow.org/api_docs/python/tf/keras/applications) works fine with TF Keras's, [Sequential](https://www.tensorflow.org/api_docs/python/tf/keras/Sequential) or [Functional](https://www.tensorflow.org/guide/keras/functional) Models?

Thanks!

**Comment by sgasse** [comment]
@rmothukuru  sry I am not working on the project that I worked on when I initially reported this issue.

The main issue back then was that Estimators handle multiple sessions (for parallel training and eval) while `tf.keras.applications` were backed by a single session handled by the `tf.keras` backend`. I know that many layer APIs are now predominantly available/used through `tf.keras` - though I guess this is still different from using a `tf.keras.applications` module.

Has the architecture in that regard changed? E.g. would keras understand/allow to use several sessions if initialized in the model part of an estimator? @golden0080 have you maybe tested with TF2.X?

**Comment by rmothukuru** [comment]
@sgasse,
Even **`Sessions`** are deprecated in **`Tensorflow 2.x`**.
